### PR TITLE
Allow inhomogeneous Dirichlet in Step-60

### LINKED ISSUE #13034: Wrong .prm files in step-60
The `parameters.prm` files in step-60 are different w.r.t what you obtain after running the program. For instance we have https://github.com/dealii/dealii/blob/e1eec081b82e61b186f119c3611d643b1ee3a848/examples/step-60/step-60.cc#L248

while in the tutorial this is set to 7 in the .prm file. Also, if that parameter is set to 7, the following exception (concerning the discrete inf-sup) is thrown when running the TestCase 1. Notice that the **ratio** parameter is not what is reported in the tutorial.

```
DEAL::Embedded dofs: 129
DEAL::Embedding minimal diameter: 0.0110485, embedded maximal diameter: 0.0147247, **ratio: 1.33273**

----------------------------------------------------
Exception on processing: 

--------------------------------------------------------
An error occurred in line <823> of file </Users/.../examples/step-60/step-60.cc> in function
    void Step60::DistributedLagrangeProblem<1, 2>::setup_grids_and_dofs() [dim = 1, spacedim = 2]
The violated condition was: 
    embedded_space_maximal_diameter < embedding_space_minimal_diameter
Additional information: 
    The embedding grid is too refined (or the embedded grid is too
    coarse). Adjust the parameters so that the minimal grid size of the
    embedding grid is larger than the maximal grid size of the embedded
    grid.
--------------------------------------------------------

Aborting!
----------------------------------------------------
```

### PR DESCRIPTION
**Pull request by fdrmrc**
This PR fixes #13034 and allows inhomogeneous Dirichlet bcs. Indeed, with the current version, the `rhs` vector is not used and changing only the call to `VectorTools::interpolate_boundary_values()` doesn't produce the right result as the inhomogeneous constraints are not distributed to the `rhs`. Now it works even with non-constant bcs, like the one in the picture.
sin_cos

### CODE REVIEW

File: examples/step-60/doc/results.dox
Comment (luca-heltai) [inline examples/step-60/doc/results.dox:128]: Maybe I would rename this to `Embedding Dirichlet boundary conditions`

File: examples/step-60/doc/results.dox
Comment (luca-heltai) [inline examples/step-60/doc/results.dox:360]: Same here: `Embedding Dirichlet boundary conditions`

File: examples/step-60/step-60.cc
Comment (luca-heltai) [inline examples/step-60/step-60.cc:370]: embedding_dirichlet_boundary_conditions

File: examples/step-60/step-60.cc
Comment (luca-heltai) [inline examples/step-60/step-60.cc:591]: This is not necessary, since by default the function is set to "0"

File: examples/step-60/step-60.cc
Comment (luca-heltai) [inline examples/step-60/step-60.cc:856]: Shouldn't this be the `embedding_dirichlet_boundary_values` instead of `Functions::ZeroFunction<spacedim>()`?

**Review by luca-heltai — CHANGES_REQUESTED** [review]
I think you are still missing a small change in the `assemble_system` function, since you also need to pass the `embedding_rhs` to the MatrixTools::create_laplace_matrix function

File: examples/step-60/step-60.cc
Comment (fdrmrc) [reply examples/step-60/step-60.cc:370]: That's related to the value of the forcing term f, the first component of the right-hand side, which is named `embedding_rhs`, not to the bcs.

File: examples/step-60/step-60.cc
Comment (fdrmrc) [reply examples/step-60/step-60.cc:856]: Thanks Luca for the review. 

> I think you are still missing a small change in the `assemble_system` function, since you also need to pass the `embedding_rhs` to the MatrixTools::create_laplace_matrix function

Here, `embedding_value_function` is related to the first component of the right-hand side, namely `embedding_rhs`, and both are used in the call to `MatrixTools::create_laplace_matrix()`.

**Comment by fdrmrc** [comment]
Thanks Luca for the review. 

> I think you are still missing a small change in the `assemble_system` function, since you also need to pass the `embedding_rhs` to the MatrixTools::create_laplace_matrix function

Here, `embedding_value_function` is related to the first component of the right-hand side, namely `embedding_rhs`, and both are used in the call to `MatrixTools::create_laplace_matrix()`.

**Comment by luca-heltai** [comment]
Ciao Marco, welcome! Please, make sure you run "make indent" before committing. The continuous integration system is complaining that step-60 is not indented correctly. Other than this, this looks fine to me!

File: examples/step-60/step-60.cc
Comment (luca-heltai) [reply examples/step-60/step-60.cc:856]: Ah, ok. In the comment to the PR you talked about Dirichlet BCs, but your changes actually add a non homogeneous rhs, rather than non-homogeneous Dirichlet BCs.

Comment (fdrmrc) [reply examples/step-60/step-60.cc:856]: If the user doesn't change `Functions::ZeroFunction<spacedim>()`, then you're right: I only allowed a possibly non-zero rhs.

But with my changes, if the "f"  in the original problem is 0, and the user specifies a non-homogeneous Dirichlet BCs on dOmega, the call to `create_laplace_matrix()` will distribute correctly the inhomogeneous constraints to the rhs vector. Before this PR, those constraints were not distributed in this scenario. Of course everything works with a non-zero "f" too, like in the picture above.

**Comment by fdrmrc** [comment]
I noticed one last thing here:

https://github.com/dealii/dealii/blob/bc4c77646a517c123d576d5e126f17ff8c5d0831/examples/step-60/doc/results.dox#L270-L273

This is not what I observe on my machine after several tests. For instance, the default .prm file gives:
```
DEAL::Embedded dofs: 257
DEAL::Embedding minimal diameter: 0.0110485, embedded maximal diameter: 0.00736292, ratio: 0.666416
DEAL::Embedding dofs: 2429
DEAL:cg::Starting value 0.117692
DEAL:cg::Convergence step 598 value 7.68984e-13


+---------------------------------------------+------------+------------+
| Total CPU time elapsed since start          |     0.989s |            |
|                                             |            |            |
| Section                         | no. calls |  CPU time  | % of total |
+---------------------------------+-----------+------------+------------+
| Assemble coupling system        |         1 |     0.037s |       3.7% |
| Assemble system                 |         1 |    0.0877s |       8.9% |
| Output results                  |         1 |     0.116s |        12% |
| Setup coupling                  |         1 |     0.032s |       3.2% |
| Setup grids and dofs            |         1 |     0.161s |        16% |
| Solve system                    |         1 |     0.551s |        56% |
+---------------------------------+-----------+------------+------------+
```
Can I delete these lines?

**Comment by fdrmrc** [comment]
I've added the function for Dirichlet BCs. The only thing that has to be changed is the comment above about CPU times.

The indentation failure was coming from some whitespaces in `results.dox` file.

File: examples/step-60/step-60.cc
Comment (luca-heltai) [reply examples/step-60/step-60.cc:370]: If I understand this correctly, this is the rhs of the Poisson problem, right? I would not name this in the same way as the other term, simply because they have a different meaning. Maybe I would rename this `embedding_rhs_function`, and similarly for the parameter files.

**Comment by luca-heltai** [comment]
Could you address my last comment, and also change the documentation of the tutorial, including the description of the equations? We now have a `f` in the rhs, and a `u_D`  as boundary data. Otherwise the program does not do what is described by the equations at the beginning.

**Comment by fdrmrc** [comment]
Sure, thanks. Should I delete those lines about CPU times?

File: examples/step-60/step-60.cc
Comment (fdrmrc) [reply examples/step-60/step-60.cc:370]: Right, it's the rhs for Poisson. I'll follow your suggestion.

**Comment by luca-heltai** [comment]
Yes, indeed, please adjust the comment about the CPU time. A lot of things have changed in the library since the implementation of step-60. :)

**Comment by fdrmrc** [comment]
Everything should be okay now!

**Review by luca-heltai — APPROVED** [review]

**Comment by luca-heltai** [comment]
One last thing: can you squash all your commits into one, and add a file in the directory: `doc/news/changes/minor` with a short description of your change? You already have one file there, so you should now the drill. :)

I'll let someone else comment on this before merging, since you are one of my PhD students. :)

File: doc/news/changes/minor/20211224MarcoFeder
Comment (jppelteret) [inline doc/news/changes/minor/20211224MarcoFeder:3]: I think that you need a new line at the end of this file.

File: examples/step-60/doc/intro.dox
Comment (jppelteret) [inline examples/step-60/doc/intro.dox:63]: > a Dirichlet data

This doesn't read right to me. Do you mean "Dirichlet conditions", or something else?

File: examples/step-60/doc/results.dox
Comment (jppelteret) [inline examples/step-60/doc/results.dox:128]: The documentation/commentary for these new `set` functions is exactly the same as what's found elsewhere in the tutorials, right?

**Review by jppelteret — COMMENTED** [review]
Looks good to me, thanks @fdrmrc . I have just a couple of very small questions, which shouldn't take long to resolve.

File: examples/step-60/doc/intro.dox
Comment (fdrmrc) [reply examples/step-60/doc/intro.dox:63]: Yes, exactly. I could replace that with "a Dirichlet boundary condition", since that is imposed on dOmega only

File: examples/step-60/doc/results.dox
Comment (fdrmrc) [reply examples/step-60/doc/results.dox:128]: Yes

**Review by masterleinad — APPROVED** [review]
Apart from @jppelteret comments, this looks good to me.

**Comment by fdrmrc** [comment]
Thanks for the check @masterleinad. Everything should be okay now

File: examples/step-60/doc/intro.dox
Comment (jppelteret) [reply examples/step-60/doc/intro.dox]: Yes, that would be perfect thanks!

File: examples/step-60/doc/results.dox
Comment (jppelteret) [reply examples/step-60/doc/results.dox:128]: Great, thanks for the confirmation. I did a quick check and it looked identical, but I thought I'd just make sure.

Comment (jppelteret) [reply examples/step-60/doc/results.dox:128]: Ah, I see that you did this already. Awesome :-)

**Review by jppelteret — APPROVED** [review]
Thanks a lot for the nice extension @fdrmrc!

**Pull request merged**
